src/storage/btree_load.c

```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
+  if (!DB_IS_NULL (key))
+    {
+      pr_clear_value (key);
+    }
+
+start_alg:
+  if (page != NULL)
+    {
+      /* Scan has already started. key_cnt and header must be set. */
+      assert (local_header != NULL);
+      assert (*key_cnt != -1);
+      /* Check if the slot required actually fits in the page. */
+      if (*slot_id == 0 || *slot_id == *key_cnt + 1 || num_visible == 0)
+	{
+	  /* No longer fits in page, or no visible oids in current page. Need to go to the next page. */
+	  /* Get the corresponding next vpid. */
+	  next_vpid = is_desc ? local_header->prev_vpid : local_header->next_vpid;
+
+	  /* If next_vpid is NULL, it means the search has ended. */
+	  if (VPID_ISNULL (&next_vpid))
+	    {
+	      goto end;
+	    }
+
+	  /* Save the old page. */
+	  old_page = page;
+
+	  /* Set correct VPID. */
+	  *vpid = next_vpid;
+
+	  /* Init variables. */
+	  *slot_id = -1;
+	  local_header = NULL;
+	  *key_cnt = -1;
+	}
+    }
+
+  /* Scan has not started yet or a new page is required. */
+  if (page == NULL)
+    {
+      /* Get page and fix it. */
+      page = pgbuf_fix (thread_p, vpid, OLD_PAGE, PGBUF_LATCH_READ, PGBUF_UNCONDITIONAL_LATCH);
+      if (page == NULL)
+	{
+	  ASSERT_ERROR_AND_SET (ret);
+	  return ret;
+	}
+      first_scan = true;
+
+      /* Unfix the old page, if any. */
+      if (old_page != NULL)
+	{
+	  pgbuf_unfix_and_init (thread_p, old_page);
+	}
+    }
+
+  /* Check page. */
+  (void) pgbuf_check_page_ptype (thread_p, page, PAGE_BTREE);
+
+  if (local_header == NULL)
+    {
+      /* Get the header of the page. */
+      local_header = btree_get_node_header (thread_p, page);
+    }
+
+  if (*key_cnt == -1)
+    {
+      /* Get number of keys in page. */
+      *key_cnt = btree_node_number_of_keys (thread_p, page);
+    }
+
+  /* Get the number of visible oids if its the first scan of the current leaf. */
+  if (mvcc != NULL && num_visible == -1)
+    {
+      num_visible = btree_get_number_of_items (thread_p, btid, page, mvcc);
+      if (num_visible == 0)
+	{
+	  goto start_alg;
+	}
+      if (num_visible < 0)
+	{
+	  /* Error. */
+	  ret = num_visible;
+	  goto end;
+	}
+    }
+
+  adv = is_desc ? -1 : 1;
+
+  /* Check if this is the first time we try to get a value from this page. */
+  if (*slot_id == -1)
+    {
+      *slot_id = is_desc ? *key_cnt : 1;
+    }
+
+  ret = btree_get_value_from_leaf_slot (thread_p, btid, page, *slot_id, key);
+
+  *slot_id += adv;
+
+end:
+
+  *header = local_header;
+  *pg_ptr = page;
+  return ret;
+}
+
+/*
+ *  btree_get_number_of_items():-	Gets number of visible oids from a page.
+ *
+ *  thread_p(in):		    -	Thread entry.
+ *  btid(in):			    -	B-tree info.
+ *  pg_ptr(in):			    -	Page pointer.
+ *  mvcc_snapshot(in):		    -	The MVCC snapshot.
+ *
+ *  return: Number of visible items, or error code if any error occurs.	    
+ */
+
+static int
+btree_get_number_of_items (THREAD_ENTRY * thread_p, BTID_INT * btid, PAGE_PTR pg_ptr, MVCC_SNAPSHOT * mvcc_snapshot)
```

Comment:
convert btree_get_number_of_items to btree_has_any_visible. return error code and output true/false. if snapshot is nil, early out with answer true.